adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/AmqpAdapterSaslAuthenticatorFactory.java

```diff
@@ -171,46 +179,53 @@ public void process(final Handler<Boolean> completionHandler) {
                 LOG.debug("client device provided an empty list of SASL mechanisms [hostname: {}, state: {}]",
                         sasl.getHostname(), sasl.getState());
                 completionHandler.handle(Boolean.FALSE);
-            } else {
-                final String remoteMechanism = remoteMechanisms[0];
-                LOG.debug("client device wants to authenticate using SASL [mechanism: {}, host: {}, state: {}]",
-                        remoteMechanism, sasl.getHostname(), sasl.getState());
-
-                final Context currentContext = Vertx.currentContext();
-                final Future<DeviceUser> deviceAuthTracker = Future.future();
-                deviceAuthTracker.setHandler(outcome -> {
-                    if (outcome.succeeded()) {
-                        currentSpan.log("credentials verified successfully");
-                        // add span to connection so that it can be used during the
-                        // remaining connection establishment process
-                        protonConnection.attachments().set(AmqpAdapterConstants.KEY_CURRENT_SPAN, Span.class, currentSpan);
-                        final Device authenticatedDevice = outcome.result();
-                        protonConnection.attachments().set(AmqpAdapterConstants.KEY_CLIENT_DEVICE, Device.class,
-                                authenticatedDevice);
-                        succeeded = true;
-                        sasl.done(SaslOutcome.PN_SASL_OK);
-
-                    } else {
-                        TracingHelper.logError(currentSpan, outcome.cause());
-                        currentSpan.finish();
-                        LOG.debug("validation of credentials failed: " + outcome.cause().getMessage());
-                        sasl.done(SaslOutcome.PN_SASL_AUTH);
-
-                    }
-                    // invoke the completion handler on the calling context.
-                    currentContext.runOnContext(action -> completionHandler.handle(Boolean.TRUE));
-                });
-
-                final byte[] saslResponse = new byte[sasl.pending()];
-                sasl.recv(saslResponse, 0, saslResponse.length);
-
-                if (AuthenticationConstants.MECHANISM_PLAIN.equals(remoteMechanism)) {
-                    verifyPlain(saslResponse, deviceAuthTracker.completer());
-                } else if (AuthenticationConstants.MECHANISM_EXTERNAL.equals(remoteMechanism)){
-                    verifyExternal(deviceAuthTracker.completer());
+                return;
+            }
+
+            if (connectionLimitManager.isConnectionLimitExceeded()) {
+                LOG.debug("Connection limit exceeded, reject connection request");
+                completionHandler.handle(Boolean.TRUE);
```

Comment:
IMHO this is not enough to signal to the client that the SASL handshake has failed. we need to explicitly invoke `sasl.done()` with an appropriate state. In this case, `SaslOutcome.PN_SASL_TEMP` seems to be appropriate in order to signal an unrecoverable but temporary problem.